### ISSUE TITLE
Closing modal from async function leave modal open

### ISSUE THREAD

**Issue by jchatrny** [mantinedev/mantine#3536]
### What package has an issue

@mantine/modals

### Describe the bug

When closing modal from async function, modal is sometimes kept open and does not react on any user interaction. 
Just click several times on button in example to reproduce. Example is just illustrative. In our case closing is img onload, after some validations.

### What version of @mantine/hooks page do you have in package.json?

5.10.3

### If possible, please include a link to a codesandbox with the reproduced problem

https://codesandbox.io/s/zen-bas-dg3qok?file=/src/App.tsx

### Do you know how to fix the issue

No

### Are you willing to participate in fixing this issue and create a pull request with the fix

Yes

**Comment by jchatrny** [comment]
Might be related to https://github.com/mantinedev/mantine/issues/3289

**Comment by rtivital** [comment]
Use `transitionDuration={0}` – https://codesandbox.io/s/hardcore-sound-jwee8e?file=/src/App.tsx

**Comment by cyantree** [comment]
@jchatrny I think these issues are not related as this issue (#3536) doesn't use the ModalManager.

I've done some digging and nailed it down to this line:
https://github.com/mantinedev/mantine/blob/master/src/mantine-core/src/Transition/use-transition.ts#L54

Basically the hook works as the following:
When unmounting the current timeout gets cleared. However in this line an additional timeout will be created. So in case the unmounting happens to quickly this timeout still gets executed and therefore disrupts unmounting of the component.

My proposed solution is the following:
- Instead of keeping track of the current timeout in `timeoutRef`, use this ref to store a destructor method that destructs everything relevant to the current transition phase.

In this case the code block my look something like this:
```ts
      const preStateTimeout = window.setTimeout(() => {
        typeof preHandler === 'function' && preHandler();
        setStatus(shouldMount ? 'entering' : 'exiting');
      }, 10);

      const enterTransitionTimeout = window.setTimeout(() => {
        window.clearTimeout(preStateTimeout);
        typeof handler === 'function' && handler();
        setStatus(shouldMount ? 'entered' : 'exited');
      }, transitionDuration);
      timeoutRef.current = () => {
        window.clearTimeout(enterTransitionTimeout);
        window.clearTimeout(preStateTimeout);
      };
```

Cancelling the timeouts then becomes something like this:
```ts
timeoutRef.current();
```

@rtivital If it's fine for you I can create a PR for this.

**Comment by rtivital** [comment]
You are welcome to create a PR to `v6` branch

**Comment by cyantree** [comment]
I've looked into `v6` and can't reproduce the issue there anymore because the modal works different than before.

I've also checked other components that use `useTransition()` but couldn't reproduce the issue there too.

So IMHO in theory there's still an issue with this piece of code but I couldn't bring it up again.

Should I create the PR nevertheless or do we change nothing until it pops up again?
